src/corehost/cli/hostpolicy/hostpolicy.cpp

```diff
@@ -55,11 +61,49 @@ namespace
 
         return StatusCode::Success;
     }
+
+    int get_or_create_hostpolicy_context(
+        hostpolicy_init_t &hostpolicy_init,
+        const arguments_t &args,
+        bool breadcrumbs_enabled,
+        std::shared_ptr<hostpolicy_context_t> &context)
+    {
+        context = g_context_weak.lock();
+        if (context != nullptr)
+        {
+            // [TODO] Validate the current context is acceptable for this request
+            trace::info(_X("Host context has already been initialized"));
+            return StatusCode::Success;
+        }
+
+        {
+            std::lock_guard<std::mutex> lock{ g_context_lock };
+            context = g_context_weak.lock();
+            if (context != nullptr)
+            {
+                trace::info(_X("Host context has already been initialized"));
+                return StatusCode::Success;
+            }
+
+            std::unique_ptr<hostpolicy_context_t> context_local(new hostpolicy_context_t());
+            int rc = context_local->initialize(hostpolicy_init, args, breadcrumbs_enabled);
+            if (rc != StatusCode::Success)
+                return rc;
+
```

Comment:
I think weakly ordered systems will require a memory barrier here.  So that the initialization is visible before the globals are written.

Reply:
So in ARM's weakly ordered model, the memory writes from one processor do not have to be observed in the same order by another processor. We usually use thread synchronization primitives to make sure writes are thread safe.  In CoreCLR we use Locks, MemoryBarrier, and CompareExchange to make sure write are ordered.

The issue occurs in multithreaded systems, when one thread writes data to be consumed by other threads.

In this code there was a potential race on reading the shared/weak ptrs.  I wanted to make sure the writes to initialize the `first` context were visible to the other threads, before the pointer was visible (allowing the other threads to use the context.)

- A memory write barrier between the initialize and the shared ptr write would have worked. It would have guaranteed that all previous writes by the current processor were visible before subsequent writes.
- The other common approach would be to CompareExchange the global pointer.  This comes typically with the required barrier semantics.
- Also if we required the context creation to obtain a lock, it would also solve the problem, as the unlock of the writer has a memory barrier.

So I think in Aarons proposal, the first context will be fully configured before it is unlocked by a `run` or destroyed.  So the unlock code should solve any thread races.  So I suspect this issue will go away.

Reply:
Ah, thanks. I'll get rid of the tracking of an existing context for this review (since as you mentioned, we will be doing it a bit differently) and switch the getting delegate case back (for now) to creating a context locally if coreclr is not loaded.